File: chapter02/python/MonotonicQueue.py
"""
时间复杂度O(n), 空间复杂度O(k), 虽然push中有while, 但每个元素至多pop和push一次, 所以时间复杂度为O(n)
"""

from collections import deque


class MontionicQueue:
    def __init__(self):
        self.deque = deque([])

    # 队尾添加元素
    def push(self, n):
        while len(self.deque) > 0 and self.deque[-1] < n:
            self.deque.pop()
        self.deque.append(n)

    # 返回当前队列中最大元素
    def max(self):
        return self.deque[0]

    # 如果队首元素是n, 就删除它
    def pop(self, n):
        if len(self.deque) > 0 and self.deque[0] == n:
            self.deque.popleft()


def max_slinding_window(nums, k):
    dq = MontionicQueue()
    res = []
    for i in range(len(nums)):
        if i < k - 1:
            dq.push(nums[i])
        else:
            dq.push(nums[i])
            res.append(dq.max())
            dq.pop(nums[i - k + 1])
    return res


if __name__ == "__main__":
    nums = [1, 3, -1, -3, 5, 3, 6, 7]
    k = 3
    print(max_slinding_window(nums, k))

File: chapter02/cpp/MonotonicQueue.cpp
#include <iostream>
#include <vector>
#include <string>
#include <algorithm>
#include <stdlib.h>
#include <list>

using namespace std;

class MontionicQueue
{   

    private:
        list<int> pq;
    public:
        void push(int x)
        {
            while(pq.size() > 0 && pq.back() < x)
                pq.pop_back();
            pq.push_back(x);
        }
        int max()
        {
            return *pq.begin();
        }

        void pop(int x)
        {
            if(pq.front() == x)
                pq.pop_front();
        }
        int size()
        {
            return pq.size();
        }
        void show_pq()
        {
             cout<<"listOne.begin()--- listOne.end():"<<endl;
            for (list<int>::iterator i = pq.begin(); i != pq.end(); ++i)
            cout << *i << " ";
            cout << endl;
        }
};

vector<int> max_slinding_window(int* nums, int k, int l)
{
    MontionicQueue dq;
    
    vector<int> v(l-k+1, 0);

    for(int i=0; i<l; i++)
    {
        if(i < k-1)
        {
            dq.push(nums[i]);
            // dq.show_pq();
        }
            
        else
        {   
            dq.push(nums[i]);
            v[i-k+1] = dq.max();
            dq.pop(nums[i-k]);
            
        }
    }
    return v;
};

int main()
{
    int nums[] = {1, 3, -1, -3, 5, 3, 6, 7};
    int l = sizeof(nums) / sizeof(nums[0]); 
    int k = 3;
    vector<int> v = max_slinding_window(nums, k, l);
    for(int i=0; i<v.size(); i++)
        cout << v[i] << endl;
    return 0;
}
